File: CSharp/Eta/Applications/PerfTools/Common/ItemWatchlist.cs
/*|-----------------------------------------------------------------------------
 *|            This source code is provided under the Apache 2.0 license      --
 *|  and is provided AS IS with no warranty or guarantee of fit for purpose.  --
 *|                See the project's LICENSE.md for details.                  --
 *|           Copyright (C) 2022-2023 Refinitiv. All rights reserved.            --
 *|-----------------------------------------------------------------------------
 */

using System.Collections.Generic;

namespace LSEG.Eta.PerfTools.Common
{
    /// <summary>
    /// Item requests list
    /// </summary>
    public class ItemWatchlist
    {
        private List<ItemInfo> m_EntryList;
        private int m_Index;

        /// <summary>
        /// Instantiates a new item watchlist
        /// </summary>
        /// <param name="count">the start capacity of the item list</param>
        public ItemWatchlist(int count)
        {
            m_EntryList = new List<ItemInfo>(count);
            m_Index = 0;
        }

        /// <summary>
        /// Initializes the Watchlist
        /// </summary>
        public void Init() => Clear();

        /// <summary>
        /// Clears items in the watch list
        /// </summary>
        public void Clear()
        {
            m_EntryList.Clear();
            m_Index = 0;
        }

        /// <summary>
        /// Add item to the watch list
        /// </summary>
        /// <param name="itemInfo">the item info to be added</param>
        public void Add(ItemInfo itemInfo) => m_EntryList.Add(itemInfo);

        /// <summary>
        /// Get next item in the list, moving index to the beginning 
        /// if last item is returned
        /// </summary>
        /// <returns>next item in the list, null if list is empty</returns>
        public ItemInfo? GetNext()
        {
            if (m_EntryList.Count == 0)
                return null;
            ItemInfo entry = m_EntryList[m_Index];
            if (++m_Index == m_EntryList.Count)
                m_Index = 0;
            return entry;
        }

        /// <summary>
        /// Get first item in the list
        /// </summary>
        /// <returns>first item in the list, null if list is empty</returns>
        public ItemInfo? GetFront() => m_EntryList.Count == 0 ? null : m_EntryList[0];

        /// <summary>
        /// Removes first item in the list and returns it
        /// </summary>
        /// <returns>first item in the list that is removed, null if list is empty</returns>
        public ItemInfo? RemoveFront()
        {
            if (m_EntryList.Count == 0)
                return null;
            ItemInfo entry = m_EntryList[0];
            m_EntryList.RemoveAt(0);
            return entry;
        }

        /// <summary>
        /// Number of items in the watch list
        /// </summary>
        /// <returns>Number of items in the watch list</returns>
        public int Count() => m_EntryList.Count;
    }
}

File: Java/Eta/Applications/PerfTools/src/main/java/com/refinitiv/eta/perftools/common/ItemWatchlist.java
/*|-----------------------------------------------------------------------------
 *|            This source code is provided under the Apache 2.0 license
 *|  and is provided AS IS with no warranty or guarantee of fit for purpose.
 *|                See the project's LICENSE.md for details.
 *|           Copyright (C) 2019-2022 LSEG. All rights reserved.     
 *|-----------------------------------------------------------------------------
 */

package com.refinitiv.eta.perftools.common;

import java.util.ArrayList;

/**
 * Item requests list. 
 */
public class ItemWatchlist
{
    private ArrayList<ItemInfo> _entryList;   // list of entries
    int _index;      // current index for cyclic access

    /**
     * Instantiates a new item watchlist.
     *
     * @param count the count
     */
    public ItemWatchlist(int count)
    {
        _entryList = new ArrayList<ItemInfo>(count); 
        _index = 0;
    }

    /**
     * Clears items in the watch list.
     */
    public void init()
    {
        clear();
    }

    /**
     * Clears items in the watch list.
     */
    public void clear()
    {
        for (ItemInfo itemInfo : _entryList)
        {
            itemInfo.clear();
        }
        _index = 0;
    }

    /**
     * Add item to the watch list.
     *
     * @param itemInfo the item info
     */
    public void add(ItemInfo itemInfo)
    {
        _entryList.add(itemInfo);
    }

    /**
     * Get next item in the list, moving index to the beginning if last item is
     * returned.
     * 
     * @return - next item in the list,  null if list is empty..
     */
    public ItemInfo getNext()
    {
        if (_entryList.isEmpty())
            return null;

        // get the current item
        ItemInfo item = _entryList.get(_index);

        // reset to beginning
        if (++_index == _entryList.size())
        	_index = 0;

        return item;
    }
    
    /**
     * Get first item in the list.
     * 
     * @return - first item in the list, null if list is empty.
     */
    public ItemInfo getFront()
    {
        if (_entryList.isEmpty())
            return null;

        return _entryList.get(0);
    }
    
    /**
     * Removes first item in the list and returning it.
     * 
     * @return - first item in the list that is removed, null if list is empty.
     */
    public ItemInfo removeFront()
    {
        if (_entryList.isEmpty())
            return null;

        return _entryList.remove(0);
    }
    
    /**
     * Number of items in the watch list.
     * 
     * @return Number of items in the watch list.
     */
    public int count()
    {
        return _entryList.size();
    }

}